### PR TITLE
[+] Segments - Users can create segments with SQL queries

### PR DESCRIPTION
**Pull request by SeyZ**
## Pull Request checklist:

- [ ] Write an explicit title for the Pull Request
- [ ] Write changes made in the CHANGELOG.md
- [ ] Create automatic tests
- [ ] No automatic tests failures
- [ ] Test manually the implemented changes
- [ ] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code

### CODE REVIEW

File: app/services/forest_liana/resources_getter.rb [hunk 1]
```diff
@@ -19,6 +19,19 @@ def perform
         @records = @records.where(@segment.where.call())
       end
 
+      # Live Query mode
+      if @params[:query]
```
Comment (arnaudbesnier) [inline app/services/forest_liana/resources_getter.rb]: very hard to understand that it is in a context of a Live Query Segment.

File: app/services/forest_liana/query_stat_getter.rb [hunk 2]
```diff
@@ -11,7 +9,7 @@ def initialize(params)
     def perform
       raw_query = @params['query'].strip
 
-      check_query(raw_query)
+      LiveQueryChecker.check_query(raw_query)
```
Comment (arnaudbesnier) [inline app/services/forest_liana/query_stat_getter.rb]: What about?
```ruby
LiveQueryChecker.check_query(raw_query_to_check) # NOTICE: This query checker checks the query to check. 
```
😝

Seriously, `LiveQueryChecker.validate` or  `LiveQueryChecker.check` is simpler, no?

File: app/services/forest_liana/resources_getter.rb [hunk 3]
```diff
@@ -19,6 +19,19 @@ def perform
...
+        LiveQueryChecker.check_query(@params[:query])
+
+        begin
+          result = ActiveRecord::Base.connection.execute(@params[:query])
+        rescue => error
+          raise ForestLiana::LiveQueryError.new(error.message)
+        end
+
+        @records = @records.where(id: result.to_a.map { |r| r['id'] })
```
Comment (arnaudbesnier) [inline app/services/forest_liana/resources_getter.rb]: Is it really scalable? What if my SQL query returns X millions of ids?

File: app/errors/forest_liana/live_query_error.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module ForestLiana
+  class LiveQueryError < StandardError
+    attr_reader :message
+
+    def initialize(message)
+      @message = message || 'Something went wrong'
+    end
+  end
+end
```
Comment (arnaudbesnier) [inline app/errors/forest_liana/live_query_error.rb]: There is already a custom error definition `SerializeAttributeBadFormat` in `config/initializers/errors`.

Comment (arnaudbesnier) [reply app/errors/forest_liana/live_query_error.rb]: I think you should follow the existing code convention.

File: app/services/forest_liana/resources_getter.rb (hunk 1 above)
Comment (SeyZ) [reply app/services/forest_liana/resources_getter.rb]: There's a comment for this.

**Pull request merged**